### PR TITLE
hp-Adaptation

### PR DESCRIPTION
**Pull request by MayaTatarelli**
PR to incorporate the work on hp-adaptation

### CODE REVIEW

File: src/flow_solver/flow_solver_cases/naca0012.cpp [hunk 1]
```diff
@@ -66,7 +66,8 @@ template <int dim, int nstate>
 void NACA0012<dim,nstate>::set_higher_order_grid(std::shared_ptr<DGBase<dim, double>> dg) const
 {
     const std::string mesh_filename = this->all_param.flow_solver_param.input_mesh_filename+std::string(".msh");
-    std::shared_ptr<HighOrderGrid<dim,double>> naca0012_mesh = read_gmsh<dim, dim> (mesh_filename);
+    bool use_mesh_smoothing = false;
```
Comment (jbrillon) [inline src/flow_solver/flow_solver_cases/naca0012.cpp:69]:
```suggestion
    const bool use_mesh_smoothing = false;
```

File: src/mesh/gmsh_reader.cpp [hunk 2]
```diff
@@ -709,7 +709,7 @@ bool get_new_rotated_indices(const dealii::CellAccessor<dim, spacedim>& cell,
 
 template <int dim, int spacedim>
 std::shared_ptr< HighOrderGrid<dim, double> >
-read_gmsh(std::string filename, int requested_grid_order)
+read_gmsh(std::string filename, int requested_grid_order, bool use_mesh_smoothing)
```
Comment (jbrillon) [inline src/mesh/gmsh_reader.cpp:712]:
```suggestion
read_gmsh(std::string filename, int requested_grid_order, const bool use_mesh_smoothing)
```

File: src/mesh/gmsh_reader.hpp [hunk 3]
```diff
@@ -58,7 +58,7 @@ namespace PHiLiP {
       */
     template <int dim, int spacedim>
     std::shared_ptr< HighOrderGrid<dim, double> >
-    read_gmsh(std::string filename, int requested_grid_order = 0);
+    read_gmsh(std::string filename, int requested_grid_order = 0, bool use_mesh_smoothing = true);
```
Comment (jbrillon) [inline src/mesh/gmsh_reader.hpp:61]:
```suggestion
    read_gmsh(std::string filename, int requested_grid_order = 0, const bool use_mesh_smoothing = true);
```

File: src/parameters/parameters_flow_solver.cpp [hunk 4]
```diff
@@ -190,6 +196,12 @@ void FlowSolverParam::parse_parameters(dealii::ParameterHandler &prm)
         else if (flow_case_type_string == "gaussian_bump")              {flow_case_type = gaussian_bump;}
 
         poly_degree = prm.get_integer("poly_degree");
+        
+        // get max poly degree for adaptation
+        max_poly_degree_for_adaptation = prm.get_integer("max_poly_degree_for_adaptation");
+        // -- set value to poly_degree if it is the default value
+        if(max_poly_degree_for_adaptation == 0) max_poly_degree_for_adaptation = poly_degree;
```
Comment (jbrillon) [inline src/parameters/parameters_flow_solver.cpp:203]: Gaussian bump cases need to set this explicitly when they update the `poly_degree` in the loop where the parameters object is being changed then passed to `create_FlowSolver`

File: tests/integration_tests_control_files/euler_integration/naca0012/CMakeLists.txt [hunk 5]
```diff
@@ -19,6 +19,13 @@ add_test(
         WORKING_DIRECTORY ${TEST_OUTPUT_DIR}
 )
 
+configure_file(2d_euler_subsonic_naca0012_lift_based_hp_adaptation.prm  2d_euler_subsonic_naca0012_lift_based_hp_adaptation.prm  COPYONLY)
+add_test(
+        NAME INVISCID_EULER_NACA0012_HP_ADAPTATION
```
Comment (jbrillon) [inline tests/integration_tests_control_files/euler_integration/naca0012/CMakeLists.txt:24]:
```suggestion
        NAME INVISCID_SUBSONIC_NACA0012_LIFT_BASED_HP_ADAPTATION
```

**Review by jbrillon — CHANGES_REQUESTED** [review]
Great work! I've left some comments to address before approval

File: src/parameters/parameters_flow_solver.cpp (hunk 4 above)
Comment (MayaTatarelli) [reply src/parameters/parameters_flow_solver.cpp:203]: Addressed in most recent commit

**Review by jbrillon — APPROVED** [review]
LGTM

File: src/mesh/mesh_adaptation.h [hunk 6]
```diff
@@ -51,6 +51,9 @@ class MeshAdaptation
     /// Performs fixed fraction refinement based on refinement and coarsening fractions.
     void fixed_fraction_isotropic_refinement_and_coarsening();
     
+    /// Decide whether to perform h or p refinement based on a smoothness indicator.
```
Comment (PranshulThakur) [inline src/mesh/mesh_adaptation.h:54]: It would be better to add a note here about using overintegration if hp-adaptation has issues with convergence.

**Review by PranshulThakur — COMMENTED** [review]
Nice PR to introduce the functionality of choosing between h and p adaptation in PHiLiP. I have added a comment and will approve it soon.

File: src/mesh/mesh_adaptation.cpp [hunk 7]
```diff
@@ -48,15 +48,25 @@ void MeshAdaptation<dim,real,MeshType>::fixed_fraction_isotropic_refinement_and_
...
-                                                      mesh_adaptation_param->p_coarsen_fraction);
-    
-    // If a cell is flagged for both h and p adaptation, perform only p adaptation.
-    dealii::hp::Refinement::force_p_over_h(dg->dof_handler);
+//=========================================================================================================================================================
+    using MeshAdaptationTypeEnum = Parameters::MeshAdaptationParam::MeshAdaptationType;
+    MeshAdaptationTypeEnum mesh_adaptation_type = mesh_adaptation_param->mesh_adaptation_type;
+
 
+    if(mesh_adaptation_type == MeshAdaptationTypeEnum::h_adaptation){}
```
Comment (cpethrick) [inline src/mesh/mesh_adaptation.cpp:56]: Could you add a comment indicating why nothing is done if `h_adaptation` is selected?

**Review by cpethrick — COMMENTED** [review]
Great PR, the new functionality is a good addition to PHiLiP. I have left a few small comments and am happy to approve once they are addressed!

**Review by donovan97 — APPROVED** [review]
Looks good to me!

File: src/mesh/mesh_adaptation.h (hunk 6 above)
Comment (MayaTatarelli) [reply src/mesh/mesh_adaptation.h:54]: Are you sure this is the right place to leave the comment about overintegration?

File: src/mesh/mesh_adaptation.cpp [hunk 8]
```diff
@@ -48,15 +48,25 @@ void MeshAdaptation<dim,real,MeshType>::fixed_fraction_isotropic_refinement_and_
...
+    else if(mesh_adaptation_type == MeshAdaptationTypeEnum::p_adaptation){
+        // Currently, the error indicator to flag p_adaptation is the same as h_adaptation, but it will likely change in future.
```
Comment (jbrillon) [inline src/mesh/mesh_adaptation.cpp:58]: @PranshulThakur Is this still the case?

File: src/parameters/parameters_mesh_adaptation.cpp [hunk 9]
```diff
@@ -9,6 +9,19 @@ void MeshAdaptationParam::declare_parameters (dealii::ParameterHandler &prm)
...
     prm.enter_subsection("mesh adaptation");
     {
+
+        prm.declare_entry("mesh_adaptation_type", "h_adaptation",
+                          dealii::Patterns::Selection(
+                          " h_adaptation | "
+                          " p_adaptation | "
+                          " hp_adaptation "
+                          ),
+                          "Functional that we want to use. "
```
[the remark this answers is not in the document]
Comment (MayaTatarelli) [reply src/parameters/parameters_mesh_adaptation.cpp]: Fixed.

File: src/parameters/parameters_flow_solver.cpp [hunk 10]
```diff
@@ -34,6 +34,12 @@ void FlowSolverParam::declare_parameters(dealii::ParameterHandler &prm)
                           dealii::Patterns::Integer(0, dealii::Patterns::Integer::max_int_value),
                           "Polynomial order (P) of the basis functions for DG.");
 
+        prm.declare_entry("max_poly_degree_for_adaptation", "0",
+                          dealii::Patterns::Integer(0, dealii::Patterns::Integer::max_int_value),
+                          "Maxiumum possible polynomial order (P) of the basis functions for DG "
+                          "when doing adaptive simulations. Default is 0 which actually sets "
+                          "the value to poly_degree in the code, indicating a no adaptation.");
```
[the remark this answers is not in the document]
Comment (MayaTatarelli) [reply src/parameters/parameters_flow_solver.cpp]: FIxed.

File: src/mesh/mesh_adaptation.cpp (hunk 7 above)
Comment (MayaTatarelli) [reply src/mesh/mesh_adaptation.cpp]: Fixed.

**Review by cpethrick — APPROVED** [review]
Looks good, thanks for addressing my comments!

File: src/mesh/mesh_adaptation.h (hunk 6 above)
Comment (PranshulThakur) [reply src/mesh/mesh_adaptation.h:54]: Hmm, ideally we would output the message when mesh_adaptation_type is set to hp_adaptation and the residual does not converge. Maybe you could do that in post processing of the flow solver.

File: src/mesh/mesh_adaptation.cpp (hunk 8 above)
Comment (PranshulThakur) [reply src/mesh/mesh_adaptation.cpp:59]: Nope, this comment can be deleted.

File: src/mesh/mesh_adaptation.cpp [hunk 11]
```diff
@@ -48,15 +48,26 @@ void MeshAdaptation<dim,real,MeshType>::fixed_fraction_isotropic_refinement_and_
...
-    dealii::hp::Refinement::force_p_over_h(dg->dof_handler);
+//=========================================================================================================================================================
+    using MeshAdaptationTypeEnum = Parameters::MeshAdaptationParam::MeshAdaptationType;
+    MeshAdaptationTypeEnum mesh_adaptation_type = mesh_adaptation_param->mesh_adaptation_type;
+
 
+    if(mesh_adaptation_type == MeshAdaptationTypeEnum::h_adaptation){
+        // Do nothing, cells are already flagged for h-adaptation
+    } else if(mesh_adaptation_type == MeshAdaptationTypeEnum::p_adaptation){
+        // Currently, the error indicator to flag p_adaptation is the same as h_adaptation, but it will likely change in future.
```
Comment (PranshulThakur) [inline src/mesh/mesh_adaptation.cpp:59]:
```suggestion
```

File: src/mesh/mesh_adaptation.h (hunk 6 above)
Comment (jbrillon) [reply src/mesh/mesh_adaptation.h:54]: Maybe this PR should go through before #164 so that @MayaTatarelli doesn't have a bunch of merge conflicts

Comment (PranshulThakur) [reply src/mesh/mesh_adaptation.h:54]: Agreed, this PR should be merged before #164.

File: src/mesh/mesh_adaptation.cpp (hunk 8 above)
Comment (dougshidong) [reply src/mesh/mesh_adaptation.cpp:59]: Delete, resolve this convo, and merge?

Comment (jbrillon) [inline src/mesh/mesh_adaptation.cpp:59]:
```suggestion
```

**Pull request merged**